### ISSUE TITLE
Opcode 61 in Berlin is Push2

### ISSUE THREAD

**Issue by henry-hz** [Consensys/WrappedEther.dfy#13]
Dear @DavePearce,
thanks for sharing the Wrapped Ether,
the Opcode 61 in Berlin is Push2, but I found Push1:

https://github.com/Consensys/WrappedEther.dfy/blob/dc8a523ee57c0f3690dd17cdd3bf7d6967416e09/src/WrappedEtherBinary.dfy#L45

**Comment by DavePearce** [comment]
Hi Henry,

I'm not completely following what you are saying here?  If you look at the [contract bytecode](url) we see this:

```Dafny
const BYTECODE_0 : seq<u8> := [0x60, 0x60, 0x60, 0x40, 0x52, 0x60, 0x4, 0x36, 0x10, 0x60, 0xad ...
```
So, you can see the sequence `0x60 0xad` which corresponds to `Push1 0xad` on line 45.  So, that looks ok to me as far as I can see.

**Comment by DavePearce** [comment]
Ok, so looking at the [binary file](https://github.com/Consensys/WrappedEther.dfy/blob/main/bin/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2.bin#L1) from the main chain I have, the bytecode seems to be:

```
0x6060604052600436106100af
```
So, that suggests it should be `Push2 0x00af`.  My tool does often automatically optimise `Push2 0x00XX` to `Push1 0xXX` (for various reasons) so that might be why I ended up with a `Push1`.  Still, I'm not sure why its `0xad` rather than `0xaf` ... that's wierd!

I guess this is what you're trying to point out?  Ok, yup ... I will look into it (next week though).

**Comment by DavePearce** [comment]
**UPDATE:** having thought about it, I can probably guess what's happened now.  `0x00af` is a _jump destination_ in the original contract.  So, when my tool optimises `Push2` => `Push1` then it automatically adjusts the jump destination accordingly.  However, I will still look into this as I wasn't expecting this optimisation to be applied.  I want an exact match of the original bytecode ...

**Comment by henry-hz** [comment]
Hi @DavePearce , thanks for the feedbacks, yes, I just thought that having exactly the binary code reflected into the dafny code would be the best scenario for formally verify it.
**UPDATE**: I thought that having a mechanical code generator would be a great tool to verify the compiled code using dafny, something similar to what certora is doing, but with full transparency.

**Comment by DavePearce** [comment]
> I thought that having a mechanical code generator would be a great tool

Hmmmm, interesting.  Do you mean something which generates Dafny code from bytecode?  Yeah, this is basically what we are trying to do here (see this [prototype tool](https://github.com/DavePearce/DevmProofGen)).  But, there are challenges here.  Dafny struggles a lot with this low-level code, in particular the non-linear arithmetic which arises.  Did you see this [recent talk](https://www.youtube.com/watch?v=P7ZlnTIJtSk&feature=youtu.be) by the Certora people?  Its great!

**Comment by DavePearce** [comment]
Ok, so I have investigated this issue now.  In fact, its really a versioning issue.  The contract code is generated from the `DevmProofGen` which, in turn, using the [EvmIL](https://github.com/Consensys/EvmIL) tool for decoding / encoding bytecode.  EvmIL was updated in Apr to [fix the issue around reproducing exact bytecode](https://github.com/Consensys/EvmIL/issues/144), but this was not yet released as a new version.  So, I have now released a new version and updated DevmProofGen.  

However, this WrappedEth contract still remains with the slightly different bytecode.  Potentially, I can update that at some point ... though it requires a bit of faffing around.  I will raise a separate issue for that, and may process it at some point in the future.
